### PR TITLE
Underlying CoreCLR support for new string slicing overloads

### PR DESCRIPTION
**Pull request by AtsushiKan**
https://github.com/dotnet/corefx/issues/25254

These add the underlying support for the fast versions of
these extension methods.

### CODE REVIEW

**Comment by ahsonkhan** [comment]
This PR is part of https://github.com/dotnet/corefx/issues/24072, not 25254.

File: src/mscorlib/shared/System/Span.NonGeneric.cs [hunk 1]
```diff
@@ -28,13 +28,45 @@ public static class Span
...
-            {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.text);
-            }
 
             return new ReadOnlyMemory<char>(text, 0, text.Length);
         }
 
+        /// <summary>Creates a new <see cref="ReadOnlyMemory{T}"/> over the portion of the target string.</summary>
+        /// <param name="text">The target string.</param>
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="text"/> is a null reference (Nothing in Visual Basic).</exception>
```
Comment (ahsonkhan) [inline src/mscorlib/shared/System/Span.NonGeneric.cs]: Is it necessary to specify "Nothing in Visual Basic". If so, should we add that to other comments as well (for example: https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/MemoryExtensions.Portable.cs#L63). Is there a way to indicate that `Nothing` is a keyword?

File: src/mscorlib/shared/System/Span.NonGeneric.cs [hunk 2]
```diff
@@ -28,13 +28,45 @@ public static class Span
...
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> index is not in range (&lt;0 or &gt;=text.Length).
```
Comment (ahsonkhan) [inline src/mscorlib/shared/System/Span.NonGeneric.cs]: Comment vs code mismatch. `index is not in range (<0 or >=text.Length)`

The code throws if `start > text.Length`

File: src/mscorlib/shared/System/Span.NonGeneric.cs [hunk 3]
```diff
@@ -28,13 +28,45 @@ public static class Span
...
+            if (text == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.text);
+
+            if ((uint)start > (uint)text.Length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            return new ReadOnlyMemory<char>(text, start, text.Length - start);
+        }
+
+        /// <summary>Creates a new <see cref="ReadOnlyMemory{T}"/> over the portion of the target string.</summary>
```
Comment (ahsonkhan) [inline src/mscorlib/shared/System/Span.NonGeneric.cs]: We create a ReadOnlyMemory<char> not ReadOnlyMemory{T}.

File: src/mscorlib/shared/System/Span.NonGeneric.cs (hunk 1 above)
Comment (AtsushiKan) [reply src/mscorlib/shared/System/Span.NonGeneric.cs]: The phrase appears in various XML docs (probably all copy-pasted from one instance as these were.) I don't really have a dog in this either way - I'd rather that this PR not be used as the place to debate that issue, though.

Comment (ahsonkhan) [reply src/mscorlib/shared/System/Span.NonGeneric.cs]: Fair enough.

**Pull request merged**